consensus/hotstuff/committees/consensus_committee.go

```diff
@@ -0,0 +1,201 @@
+)
+
+var errSelectionNotComputed = fmt.Errorf("leader selection for epoch not yet computed")
+
+// Consensus represents the main committee for consensus nodes. The consensus
+// committee persists across epochs.
+type Consensus struct {
+	mu    sync.RWMutex
+	state protocol.ReadOnlyState // the protocol state
+	me    flow.Identifier        // the node ID of this node
+	// TODO use uint16 in leader selection impl to halve memory usage
+	// TODO delete old entries, this uses ~200kb memory/day with above optimization
+	leaders map[uint64]*leader.LeaderSelection // pre-computed leader selection for each epoch
+}
+
+func NewConsensusCommittee(state protocol.ReadOnlyState, me flow.Identifier) (*Consensus, error) {
+
+	com := &Consensus{
+		state:   state,
+		me:      me,
+		leaders: make(map[uint64]*leader.LeaderSelection),
+	}
+
+	final := state.Final()
+
+	// pre-compute leader selection for current epoch
+	current := final.Epochs().Current()
+	_, err := com.prepareLeaderSelection(current)
+	if err != nil {
+		return nil, fmt.Errorf("could not add leader for current epoch: %w", err)
+	}
+
+	// Pre-compute leader selection for previous epoch, if it exists.
+	//
+	// This ensures we always know about leader selection for at least one full
+	// epoch into the past, ensuring we are able to not only determine the leader
+	// for block proposals we receive, but also adjudicate consensus-related
+	// challenges up to one epoch into the past.
+	previous := final.Epochs().Previous()
+	_, err = previous.Counter()
+	// if there is no previous epoch, return the committee as-is
+	if errors.Is(err, protocol.ErrNoPreviousEpoch) {
+		return com, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not get previous epoch: %w", err)
+	}
+
+	_, err = com.prepareLeaderSelection(previous)
+	if err != nil {
+		return nil, fmt.Errorf("could not add leader for previous epoch: %w", err)
+	}
+
+	return com, nil
+}
+
+func (c *Consensus) Identities(blockID flow.Identifier, selector flow.IdentityFilter) (flow.IdentityList, error) {
+	return c.state.AtBlockID(blockID).Identities(filter.And(
+		filter.IsVotingConsensusCommitteeMember,
+		selector,
+	))
+}
+
+func (c *Consensus) Identity(blockID flow.Identifier, nodeID flow.Identifier) (*flow.Identity, error) {
+	identity, err := c.state.AtBlockID(blockID).Identity(nodeID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get identity for node ID %x: %w", nodeID, err)
+	}
+	if !filter.IsVotingConsensusCommitteeMember(identity) {
+		return nil, fmt.Errorf("node with ID %x is not a valid consensus committee member at block %x", nodeID, blockID)
+	}
+	return identity, nil
+}
+
+// LeaderForView returns the node ID of the leader for the given view. Returns
+// the following errors:
+//  * epoch containing the requested view has not been set up (protocol.ErrNextEpochNotSetup)
+//  * any other error indicates an unexpected internal error
+func (c *Consensus) LeaderForView(view uint64) (flow.Identifier, error) {
```

Comment:
I am wondering what happens here, if we ask to retrieve the leader for a very old Epoch, whose selection has not been computed. 

#### Suggestions:

* I think, also an `leader.InvalidViewError` could be returned, right? 
   ```suggestion
   //  * view is from an Epoch  (leader.InvalidViewError)
   func (c *Consensus) LeaderForView(view uint64) (flow.Identifier, error) {
   ```
   as we return the result of `LeaderSelection.LeaderForView` directly: https://github.com/onflow/flow-go/blob/37d8cc075af379f983eb80128bdf9e29ba8936da/consensus/hotstuff/committees/consensus_committee.go#L134
* to check for the sentinel `InvalidViewError`, it would be great if it had a method like
   ```golang
   func (e InvalidViewError) Is(other error) bool {
	   _, ok := other.(InvalidViewError)
	   return ok
   }
   ```
* Lastly, it would be great to document the `InvalidViewError` in the `Committee` API: https://github.com/onflow/flow-go/blob/7ea3685b50578eb81a6e25c1b24c3c3e7bb87745/consensus/hotstuff/committee.go#L39-L40

Reply:
Done in https://github.com/onflow/flow-go/pull/157/commits/f6ebd3585714bf52b7ee289c217d0c996dd762f5